File: src/main/java/com/gildedrose/java/Item.java
package com.gildedrose.java;

public abstract class Item {
    protected String name;
    protected int sellIn;
    protected int quality;

    public static final int MIN_QUALITY = 0;
    public static final int MAX_QUALITY = 50;

    public Item(String name, int sellIn, int quality) {
        this.name = name;
        this.sellIn = sellIn;
        this.quality = quality;
    }

    public void updateSellInAndQuality() {
        decreaseSellIn();
        updateQuality();
    }

    protected void decreaseSellIn() {
        sellIn -= 1;
    }

    protected abstract void updateQuality();

    protected void decreaseQuality() {
        if (quality > MIN_QUALITY) {
            quality -= 1;
        }
    }

    protected void increaseQuality() {
        if (quality < MAX_QUALITY) {
            quality += 1;
        }
    }

    @Override
    public String toString() {
        return this.name + ", " + this.sellIn + ", " + this.quality;
    }
}

File: src/main/kotlin/com/gildedrose/kotlin/Item.kt
package com.gildedrose.kotlin

abstract class Item(protected val name: String, protected var sellIn: Int, protected var quality: Int) {
    open fun updateSellInAndQuality() {
        decreaseSellIn()
        updateQuality()
    }

    protected open fun decreaseSellIn() {
        sellIn -= 1
    }

    protected abstract fun updateQuality()

    protected fun decreaseQuality() {
        if (quality > MIN_QUALITY) {
            quality -= 1
        }
    }

    protected fun increaseQuality() {
        if (quality < MAX_QUALITY) {
            quality += 1
        }
    }

    override fun toString(): String = "$name, $sellIn, $quality"

    companion object {
        const val MIN_QUALITY = 0
        const val MAX_QUALITY = 50
    }
}